#144 row selection and delete

joshxg opened the pull request:
Fixes #144 

- Adds a dropdown which does some things
![image](https://user-images.githubusercontent.com/35858975/132933444-f70addbb-809e-4eab-bd2e-19870c4583d5.png)
![image](https://user-images.githubusercontent.com/35858975/132933447-192425d7-97b6-456b-8926-fd5ecf3dcbf3.png)
- Wanted to align the checkbox to the right, so made some changes around that area. Extracted the row to `DataRow` (bad name?) and add an `align` into the `column` definition.

Some issues to make from here:
- Create a hook `useListData` (?) - This can encapsulate the query for some list data and mutations/actions performed on the data to tidy things up a bit.
- Extract out `HeaderRow` component
- Made me think about what is the behaviour when a row is finalised. We want to be able to select it, but we don't want to be able to delete it. Also if there's a mix of finalised/not finalised in the list of selected things. Possibly showing on the dropdown "disabled as you have finalised things and we can't delete them" 🤷

joshxg commented on packages/common/src/ui/components/inputs/Dropdown/Dropdown.tsx:
```diff
@@ -1,16 +1,92 @@
+        {props.children}
+      </ListItemText>
+    </MenuItem>
+  );
```
A little bit of extras in here. I only noticed the icons in the zeplin a little later

![image](https://user-images.githubusercontent.com/35858975/132933963-5ed945db-0e25-4678-bd9a-a7052e910ae1.png)

And then added a little bit to be able to inset the font and have a nicer mix of rows with icons and those without. 

It's a little restrictive in the font sizes.. somewhat intentionally (for consistency) and somewhat because I didn't want to spend anymore time on this..!

mark-prins replied:
I hadn't noticed the icons before, though that looks like one of the older designs 🤷‍♂️

joshxg commented on packages/common/src/ui/components/inputs/Dropdown/Dropdown.tsx:
```diff
@@ -1,16 +1,92 @@
+      <ListItemText
+        inset={props.inset}
+        primaryTypographyProps={{ sx: { fontSize: 'inherit' } }}
+        sx={{ '&.MuiListItemText-inset': { paddingLeft: '22px' } }}
```
TIL is that sometimes these class overrides work with a space between `&` and `.` and sometimes they need no space. Unsure if intentional yet, but if it is intentional - the person who decided that needs a good chatting to.

joshxg commented on packages/common/src/ui/components/inputs/Dropdown/Dropdown.tsx:
```diff
@@ -1,16 +1,92 @@
-const StyledSelect = styled(Select)({});
+const StyledSelect = styled(Select)(({ theme }) => ({
+  borderRadius: '8px',
+  width: 160,
```
Again restrictive 🤔  probably not ideal for the width and maybe should be a min width instead. Then I thought that maybe using a `size` or `width` prop which was an enum would be better and then I gave up

mark-prins replied:
wondered similar.  leave for now though - looks great and there's plenty in this PR already

joshxg commented on packages/common/src/ui/components/inputs/Dropdown/Dropdown.tsx:
```diff
@@ -1,16 +1,92 @@
+  '& .MuiSelect-icon': {
+    // If left is not explicitly defined, sometimes the icon floats to the left
+    left: 'calc(100% - 30px)',
+    color: theme.palette.primary.main,
```
This was interesting.

I had an auto-import: `import {styled} from @material-ui/system` which made me use the color like: `theme.palette['primary']['500']`. which is technically correct? I am not sure exactly which one to be using!

mark-prins replied:
I prefer the version you've used.. easier to read.
The 500 is the font weight ( i.e. normal, with bold at about 700 ) which is why leaving it off will look the same as the `['500']` version

joshxg commented on packages/common/src/ui/components/inputs/Dropdown/Dropdown.tsx:
```diff
@@ -1,16 +1,92 @@
+  [`&.${outlinedInputClasses.focused} .${outlinedInputClasses.notchedOutline}`]:
+    {
+      borderColor: theme.palette['darkGrey'],
+    },
```
Ugh! These took me forever to override, and i'm not even sure I was meant to 🙃

mark-prins replied:
very tricksy!! impressive css work there

joshxg commented on packages/common/src/ui/components/inputs/Dropdown/Dropdown.tsx:
```diff
@@ -1,16 +1,92 @@
+    borderColor: '#e4e4eb',
+  },
+  [`&:hover .${outlinedInputClasses.notchedOutline}`]: {
+    borderColor: '#e4e4eb',
```
Should be in the theme as a `palette` I think?

mark-prins replied:
yeah, ideally would be.

joshxg commented on packages/common/src/ui/components/inputs/Dropdown/Dropdown.tsx:
```diff
@@ -1,16 +1,92 @@
+    <FormControl size="small">
+      <InputLabel
+        shrink={false}
+        sx={{ color: '#8f90a6', '&.Mui-focused': { color: '#8f90a6' } }}
```
Theme? Wanted to try figure out typographys, sure can set something up to make this automatically the color.

Also: No space between `&` and `.` for this one.

joshxg commented on packages/common/src/ui/components/inputs/Dropdown/Dropdown.tsx:
```diff
@@ -1,16 +1,92 @@
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
-export const Dropdown: FC<SelectProps> = ({ sx, ...props }) => {
-  return <StyledSelect {...props} />;
+export const Dropdown: FC<DropdownProps> = ({ label, children }) => {
```
I'm not spreading props intentionally. 1) which props? For the label or the select or the form control or what? Need to add a prop for each of the components props?

I think I am leaning toward not spreading props and specific use cases for components.. (dreaming?)

mark-prins replied:
if we can, then absolutely for the specific use cases in components. that will give us consistency of the look. allowing any props will lead to all sorts of chaos

joshxg commented on packages/common/src/ui/layout/tables/RemoteDataTable.tsx:
```diff
@@ -115,54 +104,59 @@ export const RemoteDataTable = <T extends Record<string, unknown>>({
+                    >
+                      {column.render('Header')}
+                    </TableSortLabel>
+                  </TableCell>
```
Changes here were to try and be able to right align the checkbox header.

I went searching through the mui docs table examples and found this `TableSortLabel` component which flips the icon n stuff.

the `sortDirection` on the `TableCell` is apparently for blind peeps. 

will make an issue to encapsulate this elsewhere

mark-prins replied:
wicked.
I see that the icon has a tooltip 'toggle SortBy' which isn't super informative. not sure if that can be overridden

joshxg commented on packages/common/src/ui/layout/tables/RemoteDataTable.tsx:
```diff
@@ -115,54 +104,59 @@ export const RemoteDataTable = <T extends Record<string, unknown>>({
-              </TableRow>
+              <DataRow<T>
+                cells={cells}
+                values={values as T}
```
Jeez these types are really starting to get annoying!

joshxg commented on packages/mock-server/src/schema/Transaction.js:
```diff
@@ -71,6 +71,15 @@ const TransactionQueryResolvers = {
+
+    return transactions;
+  },
+
```
I should've done this is a different pr :( - deleting transaction mock and handler in the mock server

joshxg commented on packages/transactions/src/OutboundShipment/ListView/ListView.tsx:
```diff
@@ -70,19 +87,39 @@ export const OutboundShipmentListView: FC = () => {
+              const linesToDelete = tableApi?.current?.selectedRows;
+              if (linesToDelete && linesToDelete?.length > 0) {
+                mutateAsync(linesToDelete);
+                success(`Deleted ${linesToDelete?.length} invoices`)();
```
Bit weird actually? Maybe we should return 

```
const success = (message:string) => enqueueSnackbar(message, {variant: 'success'})
```

rather than the function? 

I admit I was a bit silly and got a bit confused passing `onClick={() => success("..")}`  - seems like the most common usage

mark-prins replied:
probably should - the function was a convenience. which only seems to help in that one specific scenario.

If we do keep the function, then this might make this usage clearer:

```
                const notifyDeleteSuccess = success(`Deleted ${linesToDelete?.length} invoices`);

// and then, in the delete event:
                notifyDeleteSuccess();
```

joshxg commented on types/react-table-config.d.ts:
```diff
@@ -49,11 +49,13 @@ import {
   > extends UseFiltersColumnProps<D>,
       UseGroupByColumnProps<D>,
       UseResizeColumnsColumnProps<D>,
+      Align,
```
Seems we can add anything into the columns we like which is nice. But the fields 1) have to be optional 2) need to be defined in a type explictly.

i.e.
```
...
UseResizeColumnsColumnProps<D>,
{ align: 'left' | 'center' | 'right' },
...
```

didn't work hehe

mark-prins commented:
there might be a missing commit. the DataRow is throwing `cannot read property 'align' of undefined` in the test and storybook

joshxg commented:
Ugh sorry about that- pushed

mark-prins commented on packages/mocks/src/handlers.ts:
```diff
@@ -2,7 +2,7 @@ import { graphql } from 'msw';
 import faker from 'faker';
 
-const TransactionData = Array.from({ length: 500 }).map((_, i) => ({
+const TransactionData = Array.from({ length: 10 }).map((_, i) => ({
```
you don't like my big data set?

mark-prins reviewed (commented):
that is awesome!
Just adding as a comment, so I can refresh..

mark-prins reviewed (approved):
magic!

joshxg commented on packages/common/src/intl/locales/pt.json:
```diff
@@ -26,6 +26,8 @@
   "button.try-again": "Try again",
   "button.go-back": "Go back",
+  "button.delete-lines": "Delete selected lines",
+  "button.export-to-csv": "Export to CSV",
```
I've also been meaning to look through how the package can extract the strings from the code base also

joshxg commented on packages/common/src/ui/components/inputs/Dropdown/Dropdown.test.tsx:
```diff
@@ -0,0 +1,60 @@
+import { TestingProvider } from '../../../../utils';
+
+describe('Dropdown', () => {
+  it('Renders the dropdown item children when the dropdown is clicked', () => {
```
I am not getting this one.

When I remove 

```
act(() => {
      userEvent.click(button);
    });
```

The test fails

joshxg commented:
Renamed `Dropdown` -> `DropdownMenu` and put it in a `menus` folder (🤔 )

Reverted the handler reduction in data 😁  (accidental!!) and fixed the test description. Unsure about the flakey test?
